platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/isolated/FetchCustomModelInParallel.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.configurationcache.isolated;
+
+import org.gradle.configurationcache.fixtures.SomeToolingModel;
+import org.gradle.tooling.BuildAction;
+import org.gradle.tooling.BuildController;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class FetchCustomModelInParallel implements BuildAction<List<SomeToolingModel>> {
+
+    private final String blockingHttpServerUrl;
+
+    public FetchCustomModelInParallel(String blockingHttpServerUrl) {
+        this.blockingHttpServerUrl = blockingHttpServerUrl;
+    }
+
+    @Override
+    public List<SomeToolingModel> execute(BuildController controller) {
+        List<FetchModelForProject> actions = new ArrayList<>();
+        for (int i = 0; i < 3; i++) {
+            actions.add(new FetchModelForProject(blockingHttpServerUrl + "/nested-" + (i + 1)));
+        }
+
+        List<SomeToolingModel> models = controller.run(actions);
+        awaitBlockingHttpServer(blockingHttpServerUrl + "/finished");
+        return models;
+    }
+
+    private static class FetchModelForProject implements BuildAction<SomeToolingModel> {
+
+        private final String url;
+
+        private FetchModelForProject(String url) {
+            this.url = url;
+        }
+
+        @Override
+        public SomeToolingModel execute(BuildController controller) {
+            awaitBlockingHttpServer(url); // Ensure each nested action is running in parallel
+            return controller.findModel(SomeToolingModel.class);
+        }
+    }
+
+    // CHECKSTYLE:OFF
+    private static void awaitBlockingHttpServer(String url0) {
```

Comment:
That is what I initially wanted. However, the models are never built in parallel in this case due to a project lock:
https://github.com/gradle/gradle/blob/a3dc91fa697b5de5fd9c2f4ce13e1fbda561fa26/subprojects/core/src/main/java/org/gradle/tooling/provider/model/internal/DefaultToolingModelBuilderRegistry.java#L96

So I opted for blocking in the top-level build action, to make sure the models are _requested_ in parallel and to block as close to the model builders as possible. And the `sleep` is there to make sure the `creator()` call in the `ProjectStateStore` is taking long enough to observe multiple threads reaching that point.

In summary, the test is not about concurrent creation of models, but rather about duplicate re-creation in case of concurrent requests.

Reply:
After taking another look, I think having the blocking server there does not actually make the test any better.

We have separate tests that ensure nested requests run in parallel, so we don't need to block there. What we do need is to make sure the first creator of the model runs long enough so that the second request catches up. Then we validate that despite this, we have created only a single model (the builder executed once).

I simplified the test.